src/main/java/com/commercetools/sync/products/service/impl/ProductReferenceTransformServiceImpl.java

```diff
@@ -0,0 +1,496 @@
+package com.commercetools.sync.products.service.impl;
+
+import static com.commercetools.sync.commons.utils.ResourceIdentifierUtils.REFERENCE_ID_FIELD;
+import static com.commercetools.sync.commons.utils.ResourceIdentifierUtils.REFERENCE_TYPE_ID_FIELD;
+import static com.commercetools.sync.products.utils.ProductReferenceResolutionUtils.mapToProductDrafts;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+import com.commercetools.sync.commons.models.GraphQlQueryResources;
+import com.commercetools.sync.commons.models.ResourceIdsGraphQlRequest;
+import com.commercetools.sync.commons.utils.ChunkUtils;
+import com.commercetools.sync.customobjects.helpers.CustomObjectCompositeIdentifier;
+import com.commercetools.sync.products.service.ProductReferenceTransformService;
+import com.commercetools.sync.services.impl.BaseTransformServiceImpl;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.sphere.sdk.categories.Category;
+import io.sphere.sdk.client.SphereClient;
+import io.sphere.sdk.customobjects.CustomObject;
+import io.sphere.sdk.customobjects.queries.CustomObjectQuery;
+import io.sphere.sdk.models.Asset;
+import io.sphere.sdk.models.Reference;
+import io.sphere.sdk.products.AttributeContainer;
+import io.sphere.sdk.products.Price;
+import io.sphere.sdk.products.Product;
+import io.sphere.sdk.products.ProductDraft;
+import io.sphere.sdk.products.ProductLike;
+import io.sphere.sdk.products.ProductVariant;
+import io.sphere.sdk.products.attributes.Attribute;
+import io.sphere.sdk.producttypes.ProductType;
+import io.sphere.sdk.types.CustomFields;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ProductReferenceTransformServiceImpl extends BaseTransformServiceImpl
+    implements ProductReferenceTransformService {
+
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(ProductReferenceTransformServiceImpl.class);
+  private static final String WITH_IRRESOLVABLE_REFS_ERROR_MSG =
+      "The product with id '%s' on the source project ('%s') will "
+          + "not be synced because it has the following reference attribute(s): %n"
+          + "%s.%nThese references are either pointing to a non-existent resource or to an existing one but with a blank key. "
+          + "Please make sure these referenced resources are existing and have non-blank (i.e. non-null and non-empty) keys.";
+
+  public ProductReferenceTransformServiceImpl(
```

Comment:
What about providing a default map backed by caffeeine as a reference? So the library users would use it without investing in their best practices?

Reply:
Do you mean to provide the default map when the map is null ?  Or just update the javadocs about the default cache Caffeine for library users to implement it with example?

Reply:
I'm asking giving a reference class for to use by default on our snippets, also document for custom use cases, and invalidation scenarios for our clients (in the doc snippets)

For instance we would have a class called `DefaultTransformServiceCache` (or you name it) which is documented as an example in the docs and used by integration tests.

Class is backed by caffeine cache with LRU cache eviction strategy, which is taking the size and documents the best practices for the caching needs. It could also have additional utility methods like invalidateAll method, or others.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/697/commits/8a458e56c6e4a63b346644e26f27b6cf44057a5d
I have added Default cache using Caffeine and this will be documented as part of documentation task.

Reply:
First it's not something related to product package, maybe this should be part of the utils, also instead of explaining it, I pushed a commit to another branch, which I meant in my comment: https://github.com/commercetools/commercetools-sync-java/commit/c258f2cbe171378379d6fbc0ef44659a84f64d54

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/697/commits/840a2dea62f16b192af58dbc402883551f0a39f5 merged into this branch using cherry-pick